examples/tutorial_client_events.c

```diff
@@ -0,0 +1,116 @@
+
+#include <stdio.h>
+#include <signal.h>
+#include "open62541.h"
+
+#ifdef UA_ENABLE_SUBSCRIPTIONS
+static void
+handler_events(UA_UInt32 monId, size_t nEventFields, const UA_Variant *eventFields, const void *context) {
+    printf("got an event! %d, %zu\n", monId, nEventFields);
+
+    for(size_t i = 0; i < nEventFields; ++i) {
+        printf("%s\n", eventFields[i].type->typeName);
+        if (UA_Variant_hasScalarType(&eventFields[i], &UA_TYPES[UA_TYPES_BYTESTRING])) {
+            UA_ByteString *string = (UA_ByteString *)eventFields[i].data;
+            printf("'%.*s'\n", (int)string->length, string->data);
+            for (size_t u = 0; u < string->length; ++u)
+                printf("%x", string->data[u]);
+            printf("\n");
+        } else if (UA_Variant_hasScalarType(&eventFields[i], &UA_TYPES[UA_TYPES_LOCALIZEDTEXT])) {
+            UA_LocalizedText *lt = (UA_LocalizedText *)eventFields[i].data;
+            printf("'%.*s'\n", (int)lt->text.length, lt->text.data);
+            printf("size: %zu\n", lt->text.length);
+        }
+    }
+}
+
+static UA_Boolean running = true;
+static void stopHandler(int sig) {
+    UA_LOG_INFO(UA_Log_Stdout, UA_LOGCATEGORY_USERLAND, "received ctrl-c");
+    running = false;
+}
+#endif
+
+int main(int argc, char *argv[]) {
+    signal(SIGINT, stopHandler);
+    signal(SIGTERM, stopHandler);
+
+    UA_Client *client = UA_Client_new(UA_ClientConfig_default);
+
+    UA_StatusCode retval = UA_Client_connect(client, "opc.tcp://uademo.prosysopc.com:53530/OPCUA/SimulationServer");
+    if(retval != UA_STATUSCODE_GOOD) {
+        UA_Client_delete(client);
+        return (int)retval;
+    }
+
+#ifdef UA_ENABLE_SUBSCRIPTIONS
+    /* Create a subscription */
+    UA_UInt32 subId = 0;
+    retval = UA_Client_Subscriptions_new(client, UA_SubscriptionSettings_default, &subId);
+    if(!subId) {
+        UA_Client_disconnect(client);
+        UA_Client_delete(client);
+        return (int)retval;
+    }
+    printf("Create subscription succeeded, id %u\n", subId);
+
+    /* Add a MonitoredItem */
+    UA_NodeId monitorThis = UA_NODEID_NUMERIC(0, 2253); // Root->Objects->Server
+    UA_UInt32 monId = 0;
+
+
+    // creating a single element array for the select clause of length 1
+    const size_t nSelectClauses = 1;
+    UA_SimpleAttributeOperand *selectClauses = (UA_SimpleAttributeOperand *)UA_Array_new(nSelectClauses, &UA_TYPES[UA_TYPES_SIMPLEATTRIBUTEOPERAND]);
+    if(!selectClauses){
+        UA_Client_Subscriptions_remove(client, subId);
+        UA_Client_disconnect(client);
+        UA_Client_delete(client);
+        return (int)UA_STATUSCODE_BADOUTOFMEMORY;
+    }
+
+    for(size_t i =0; i<nSelectClauses; ++i) {
+        UA_SimpleAttributeOperand_init(&selectClauses[i]);
+    }
+
+    selectClauses[0].typeDefinitionId = UA_NODEID_NUMERIC(0, 2041) ; // BaseEventType
+    selectClauses[0].browsePathSize = 1;
+    //selectClauses[0].browsePathSize = 2; // crashes
```

Comment:
If I set a single browsePath everything work fine. 

I've set a break point in my eventhandler and print out the datastructures.

`
gdb ./tutorial_client_events 
Reading symbols from ./tutorial_client_events...done.
(gdb) b tutorial_client_events.c:14
Breakpoint 1 at 0x40147d: file /home/fmee/projects/foss/open62541/examples/tutorial_client_events.c, line 14.
(gdb) r
Starting program: /home/fmee/projects/foss/build-open62541-Desktop-Default/bin/examples/tutorial_client_events 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Create subscription succeeded, id 356
Monitoring 'Root->Objects->Server', id 356
got an event! 1, 1

Breakpoint 1, handler_events (monId=1, nEventFields=1, eventFields=0x655af0, context=0x0) at /home/fmee/projects/foss/open62541/examples/tutorial_client_events.c:14
14              printf("%s\n", eventFields[i].type->typeName);
(gdb) p eventFields[0]
$1 = {type = 0x64ac00 <UA_TYPES+960>, storageType = UA_VARIANT_DATA, arrayLength = 0, data = 0x655b30, arrayDimensionsSize = 0, arrayDimensions = 0x0}
(gdb) p eventFields[0]->type
$2 = (const UA_DataType *) 0x64ac00 <UA_TYPES+960>
(gdb) p eventFields[0].type
$3 = (const UA_DataType *) 0x64ac00 <UA_TYPES+960>
(gdb) p eventFields[0].type->typeName 
$4 = 0x4392fe "LocalizedText"
(gdb) quit
`
This looks good.

Next the bad case. I set two browsePath elements (Severity and Message).

`
gdb ./tutorial_client_events 
Reading symbols from ./tutorial_client_events...done.
(gdb) b tutorial_client_events.c:14
Breakpoint 1 at 0x40147d: file /home/fmee/projects/foss/open62541/examples/tutorial_client_events.c, line 14.
(gdb) r
Starting program: /home/fmee/projects/foss/build-open62541-Desktop-Default/bin/examples/tutorial_client_events 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Create subscription succeeded, id 357
Monitoring 'Root->Objects->Server', id 357
got an event! 1, 1

Breakpoint 1, handler_events (monId=1, nEventFields=1, eventFields=0x655b20, context=0x6ab9fde1077c2a19) at /home/fmee/projects/foss/open62541/examples/tutorial_client_events.c:14
14              printf("%s\n", eventFields[i].type->typeName);
(gdb) p eventFields[0]
$1 = {type = 0x0, storageType = UA_VARIANT_DATA, arrayLength = 0, data = 0x0, arrayDimensionsSize = 0, arrayDimensions = 0x0}
`

First observation: nEventFields is still one. I would have expected two.
Printing the variant shows that empty (all fields are 0).

The crash was my code printing out the type->typeName unconditionally.
Still I don't have an idea what I get this strange result.

Any ideas? How to debug this?